Mark `WorkingBeatmap.GetSkin()` as abstract instead of using `DefaultSkin`

frenzibyte opened the pull request:
Marks `WorkingBeatmap.GetSkin()` as abstract and locally implements for all usages instead of using `DefaultSkin` which may potentially override a parent skin's components while it shouldn't.

frenzibyte committed:
Move beatmap skin info creation to static method at `IBeatmapSkin`

frenzibyte committed:
Separate default beatmap skins from `DefaultSkin`

peppy commented:
What does the fix? Can there be test coverage of intention here?

frenzibyte commented:
This can't have test coverage on its own, the only thing you could test this with is `TestSceneCatchPlayerLegacySkin` on #12826, because it returns `Source.GetDrawableComponent(component)` on `CatchLegacySkinTransformer`, which goes through the beatmap skin (while returning `null` as previously doesn't, and I'm not sure if it's intended or not, sounds like a bug)

bdach commented:
Is there anything blocking setting up a fake/mock skin for a test that uses `Source.GetDrawableComponent()` in the way you want it to in that other PR?

I really would like to see a defined problem statement and an isolated concrete test case. Because as it is, it looks just like chasing tangential changes multiple PRs deep to a brain-breaking degree.

frenzibyte commented:
I wasn't really sure about using code that doesn't even sound right for testing this, but sure can do it.

frenzibyte commented:
Making a legacy skin transformer to return `Source.GetDrawableComponent(component)` instead of `null`, it's a "transformer", so it shouldn't matter which is returned, isn't that right?

Either way, I will change `CatchLegacySkinTransformer` to return `Source.GetDrawableComponent()` instead (can't achieve the same in any other way otherwise, unless I go ahead and write a custom ruleset backed test scene, which I'm not sure about that). 

I was just afraid of causing even more confusion with doing this, but at least it wouldn't have been as confusing as a PR that doesn't have an isolated test case for what it's for.

bdach commented:
> unless I go ahead and write a custom ruleset backed test scene, which I'm not sure about that

I would prefer this. Making unrelated changes in classes used in other actual non-test contexts for a test is wrong.

If you're not sure if the pattern you want to use is correct, make it a gist at first instead of pushing, and maybe we'll be able to make a judgement.

frenzibyte committed:
Add failing test case

frenzibyte commented:
Added a test case on the same test scene I had this on, was planning to put it on `LegacySkinPlayerTestScene` but can't happen due to the weird behaviour of skin transformers, to be resolved separately.

frenzibyte committed:
Revert "Add failing test case"

This reverts commit ec89a149dd752d242e1675a4462c8af4e360eb3e.

frenzibyte committed:
Add extensible test scene for beatmap skins fallback instead

frenzibyte commented:
I've reverted that test case and instead wrote a separate extensible test scene for this, as there are changes that I want to separate from #12826 and this test scene would be the right place for testing it.

frenzibyte commented on osu.Game.Tests/Visual/Gameplay/TestSceneBeatmapSkinFallbacks.cs:
```diff
@@ -0,0 +1,129 @@
+                var actualInfo = Player.ChildrenOfType<SkinnableTargetContainer>().First(s => s.Target == target)
+                                       .ChildrenOfType<SkinnableTargetComponentsContainer>().Single().CreateSkinnableInfo();
+
+                return almostEqual(actualInfo, expectedInfo, 2f);
```
I've sneaked in a `2f` position tolerance as apparently, for some reason, the position of the accuracy counter in the legacy HUD components set may have a very slightly different calculated Y value.

I haven't deeply checked why this is the case yet, currently thinking there's some sort of weird offset between the expected target components container, and the actual target components container, will dive further into this in a bit.

frenzibyte replied:
I've improved this, it actually just needed `0.001f` tolerance instead, not sure why it was failing for me before (probably something else I've mixed this up with)

I was initially going to put the equality logic in `SkinnableInfo` instead via `IEquatable`, but since this doesn't check for all properties to be *exactly* matching, I've just left it as-is.

frenzibyte committed:
Merge branch 'master' into beatmap-skin

frenzibyte committed:
Extract assert step addition out of assertion method

frenzibyte committed:
Improve components assertion logic

frenzibyte committed:
Remove unrequired null conditional

frenzibyte committed:
Remove unused using directive

peppy commented on osu.Game/Beatmaps/WorkingBeatmap.cs:
```diff
@@ -324,7 +324,7 @@ public Track Track
         public ISkin Skin => skin.Value;
 
-        protected virtual ISkin GetSkin() => new DefaultSkin(null);
+        protected virtual ISkin GetSkin() => new BeatmapSkin(BeatmapInfo);
```
I mentioned this on discord, but why not just make this `abstract`? Correct me if I'm wrong, but this new class you've made is *only used in tests*, at which point it should just be part of the `TestWorkingBeatmap` internals.

frenzibyte replied:
Good point, will do.

frenzibyte replied:
I've done this, and implemented it in all `WorkingBeatmap`s based on whether they throw `NotImplementedException` in other overrides or not, check if it's good.

I've also intentionally left the test scene although it doesn't have any test cases as it's needed in another PR (#12875), not worth going through the hassle of moving this from here to there.

peppy replied:
I've just removed this completely for now. A test scene is not the place to fix this issue (as it also affects every other usage of the `Transformer` logic).

frenzibyte replied:
Good choice, I'll get a PR ready for it then, should be pretty simple to understand here.

frenzibyte replied:
I've instead pushed it directly to #12875 since they both rely on each other, #12875 requires the transformers fix for its test scene to not false-pass, and the transformers fix requires #12875 because the bug will become visible and affect users if #12875 isn't merged yet.

frenzibyte committed:
Mark `WorkingBeatmap.GetSkin()` as abstract

frenzibyte committed:
Return empty skin for `GetSkin()` in `TestWorkingBeatmap`

frenzibyte committed:
Implement `GetSkin()` for other working beatmaps

frenzibyte committed:
Remove no longer necessary test case

peppy commented on osu.Game/Skinning/BeatmapSkin.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <remarks>
+    /// This should be removed once <see cref="Skin"/> becomes instantiable or a new skin type for osu!lazer beatmaps is defined.
+    /// </remarks>
+    public class BeatmapSkin : Skin
```
Not used?

peppy commented on osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs:
```diff
@@ -36,10 +42,23 @@ public TestWorkingBeatmap(IBeatmap beatmap, Storyboard storyboard = null, AudioM
 
         protected override Storyboard GetStoryboard() => storyboard ?? base.GetStoryboard();
 
+        protected override ISkin GetSkin() => new EmptySkin();
```
Is there a reason this is empty and not null?

frenzibyte committed:
Remove no longer used `BeatmapSkin`

frenzibyte committed:
Replace unnecessary empty skin implementation with null

peppy committed:
Merge branch 'master' into beatmap-skin

peppy commented on osu.Game/Skinning/BeatmapSkinExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public static class BeatmapSkinExtensions
+    {
+        public static SkinInfo CreateSkinInfo(BeatmapInfo beatmap) => new SkinInfo
```
What's the intention of this? Is it used somewhere else? Even in the follow-up PR it's only called in one spot.

frenzibyte replied:
No longer necessary now, missed it as well.

peppy commented:
The test contains no actual steps.. i don't

peppy commented on osu.Game.Tests/Visual/Gameplay/TestSceneBeatmapSkinFallbacks.cs:
```diff
@@ -0,0 +1,135 @@
+                    if (drawable != null)
+                        return drawable;
+
+                    // this isn't really supposed to make a difference from returning null,
+                    // but it appears it does, returning null will skip over falling back to beatmap skin,
+                    // while calling Source.GetDrawableComponent() doesn't.
```
I don't get this. Are you saying this is a bug? `OsuRulesetTransformer` does return `null` - are you suggesting that it is never going to use the beatmap skin as a result? Because that's how this reads.

https://github.com/ppy/osu/blob/004798d61d44d88114acfd7c2861378261eab107/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs#L120

frenzibyte replied:
That's correct, if `LegacyBeatmapSkin.GetDrawableComponent()` implements something for a certain `component` and the transformer returns `null` on that `component`, then the implementation in the beatmap skin will be completely ignored.

And this is currently a hidden bug, until beatmap skins can have skinned HUD components, in which case they'll be silently ignored and the bug would become visible.

And now that I think about it, it sounds like the transformers should just return `Source.GetDrawableComponent` if it's not handling it, while allowing to return `null` if there are no skinned textures for the component (in which case it doesn't make sense to call the source).

peppy replied:
The transformer's responsibility is to decide what should or should not be seen in the usage scenario. If the correct behaviour is that "unknowns should be provided to the consumer" then yes, the default `null` should be replaced with a fallback call to the transformation source.

peppy replied:
Which is to say, your inline comment reads very badly. Try rewording it.

And maybe it should explain why we're doing that here rather than just accepting what `base.GetDrawableComponent` is giving us?

frenzibyte commented:
> The test contains no actual steps.. i don't

That's intentional, there's nothing that would make sense to test here, but I've left the test scene for another PR in the series that needs it, as I've already mentioned in the conversation that requested the change (https://github.com/ppy/osu/pull/12862#discussion_r637085701).

frenzibyte committed:
Revert "Move beatmap skin info creation to static method at `IBeatmapSkin`"

This reverts commit 9806d94b745348e7c44af36cc93b76401234d225.

peppy committed:
Merge branch 'master' into beatmap-skin

peppy committed:
Remove misleading comment and misplaced bug fix

This shouldn't be fixed in a test scene; the underlying issue should be
fixed in actual game code.

peppy committed:
Merge branch 'master' into beatmap-skin
